### PR TITLE
MAINT, DOC: make np._from_dlpack public

### LINKED ISSUE #20743: ENH: Make `from_dlpack()` a public API under the main module
### Proposed new feature or change:

Change `np._from_dlpack` to `np.from_dlpack`, see the discussion starting at https://github.com/numpy/numpy/pull/19083#issuecomment-1004208093.

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
fixes gh-20743

This PR:

- Makes `np._from_dlpack` public.
- Adds docs for the DLPack protocol in the user guide section "Interpolability with NumPy"

cc @seberg @mattip @leofang 

@melissawm this might interest you :)

### CODE REVIEW

File: numpy/core/_add_newdocs.py [hunk 1]
```diff
@@ -1573,17 +1573,37 @@
...
     Create a NumPy array from an object implementing the ``__dlpack__``
-    protocol.
+    protocol. Generally, the returned NumPy array is a read-only view
+    of the input object. See [1]_ and [2]_ for more details.
 
-    See Also
+    Parameters
+    ----------
+    x : object
+        A Python object that implements the ``__dlpack__`` method.
```
Comment (tirthasheshpatel) [inline numpy/core/_add_newdocs.py:1587]: The spec mandates both `__dlpack__` and `__dlpack_device__` but NumPy just uses the DLTensor to retrieve the device info. So, practically `__dlpack_device__` isn't required to import arrays in NumPy but theoretically, it is. We might want to change it that way then:

```suggestion
        A Python object that implements the ``__dlpack__`` and
        ``__dlpack_device__`` methods.
```

**Comment by charris** [comment]
Won't this break backwards compatibility? Perhaps we should keep both names for a bit while deprecating the underscored version.

**Comment by tirthasheshpatel** [comment]
> Won't this break backwards compatibility?

I didn't expect backward compatibility would be a concern for private methods (_from_dlpack wasn't even documented before). The reason we kept `_from_dlpack` private was that the protocol wasn't formally documented anywhere and there were a few bugs. Do you think it's OK not to worry about backward compatibility given the above reasons? If not, I can keep both the versions :)

File: numpy/core/_add_newdocs.py (hunk 1 above)
Comment (leofang) [reply numpy/core/_add_newdocs.py:1587]: Both are required, otherwise there will be issues when consumed by downstream.

Comment (tirthasheshpatel) [reply numpy/core/_add_newdocs.py]: Added to the docs, thanks @leofang!

File: doc/source/user/basics.interoperability.rst [hunk 2]
```diff
@@ -395,6 +422,75 @@ See `the Dask array documentation
...
+ >>> x_torch = torch.from_dlpack(x_np)
+
+Read-only arrays cannot be exported:
+
+ >>> x_np = np.arange(5)
+ >>> x_np.flags.writeable = False
+ >>> torch.from_dlpack(x_np)
+ Traceback (most recent call last):
+   File "<stdin>", line 1, in <module>
+   File "/home/tirthasheshpatel/tensor_libs/virtualenvs/numpy-dev/lib/python3.9/site-packages/torch/utils/dlpack.py", line 63, in from_dlpack
```
Comment (mattip) [inline doc/source/user/basics.interoperability.rst:490]: You can use ellipsis and then add the docutils directive `# doctest: +ELLIPSIS` to the code

```suggestion
   File ".../site-packages/torch/utils/dlpack.py", line 63, in from_dlpack
```

File: numpy/core/_add_newdocs.py [hunk 3]
```diff
@@ -1573,17 +1573,38 @@
...
+    -------
+    out : ndarray
+
+    References
+    ----------
+    .. [1] Array API documentation,
+       https://data-apis.org/array-api/latest/design_topics/data_interchange.html#syntax-for-data-interchange-with-dlpack
+
+    .. [2] Python specification for DLPack,
+       https://dmlc.github.io/dlpack/latest/python_spec.html
```
Comment (mattip) [inline numpy/core/_add_newdocs.py:1600]: These should be kept as full links (not intersphinx) so people using help() will still see the info

File: doc/source/user/basics.interoperability.rst [hunk 4]
```diff
@@ -395,6 +422,75 @@ See `the Dask array documentation
 and the `scope of Dask arrays interoperability with NumPy arrays
 <https://docs.dask.org/en/stable/array.html#scope>`__ for details.
 
+Example: DLPack
+~~~~~~~~~~~~~~~
+
+Several Python data science libraries implement the ``__dlpack__`` protocol.
+Among them are PyTorch_ and CuPy_. A full list of libraries that implement
+this protocol can be found on `this page of DLPack
+documentation <https://dmlc.github.io/dlpack/latest/#purpose>`_.
```
[the remark this answers is not in the document]
Comment (mattip) [reply doc/source/user/basics.interoperability.rst:431]: Yup. The documentation has a https://dmlc.github.io/dlpack/latest/objects.inv file.

File: doc/source/user/basics.interoperability.rst [hunk 5]
```diff
@@ -395,6 +422,75 @@ See `the Dask array documentation
...
+Convert a PyTorch CPU tensor to NumPy array:
+
+ >>> import torch
+ >>> x_torch = torch.arange(5)
+ >>> x_torch
+ tensor([0, 1, 2, 3, 4])
+ >>> x_np = np.from_dlpack(x_torch)
+ >>> x_np
+ array([0, 1, 2, 3, 4])
+ >>> # note that x_np is a view of x_torch
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply doc/source/user/basics.interoperability.rst:442]: Might be distracting. It'd be good to hide/abstract most implementation details like the base containing a capsule named "numpy_dltensor".

File: doc/source/user/basics.interoperability.rst [hunk 6]
```diff
@@ -395,6 +422,75 @@ See `the Dask array documentation
...
+ >>> x_torch[1] = 100
+ >>> x_torch
+ tensor([  0, 100,   2,   3,   4])
+ >>> x_np
+ array([  0, 100,   2,   3,   4])
+
+The imported arrays are read-only so writing or operating in-place will fail:
+
+ >>> x_np[1] = 1
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply doc/source/user/basics.interoperability.rst:451]: Good idea! I will add that.

File: doc/source/user/basics.interoperability.rst [hunk 7]
```diff
@@ -55,6 +55,14 @@ describes its memory layout and NumPy does everything else (zero-copy if
 possible). If that's not possible, the object itself is responsible for
 returning a ``ndarray`` from ``__array__()``.
 
+DLPack_ is yet another protocol to convert foriegn objects to NumPy arrays in
+a language and device agnostic manner. NumPy doesn't implicitly convert objects
+to ndarrays using DLPack. It provides the function `numpy.from_dlpack` that
+accepts any object implementing the ``__dlpack__`` method and outputs a NumPy
+ndarray (which is generally a view of the input object's data buffer). The
+`Python specification for DLPack <https://dmlc.github.io/dlpack/latest/python_spec.html>`_
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply doc/source/user/basics.interoperability.rst]: Done.

File: doc/source/user/basics.interoperability.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply doc/source/user/basics.interoperability.rst]: Done.

File: doc/source/user/basics.interoperability.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply doc/source/user/basics.interoperability.rst:454]: Done.

File: doc/source/user/basics.interoperability.rst [hunk 8]
```diff
@@ -395,6 +422,75 @@ See `the Dask array documentation
...
+ array([  0, 100,   2,   3,   4])
+
+The imported arrays are read-only so writing or operating in-place will fail:
+
+ >>> x_np[1] = 1
+ Traceback (most recent call last):
+   File "<stdin>", line 1, in <module>
+ ValueError: assignment destination is read-only
+
+A copy must be created in order to operate on the imported arrays in-place:
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply doc/source/user/basics.interoperability.rst]: Added.

File: doc/source/user/basics.interoperability.rst (hunk 2 above)
Comment (tirthasheshpatel) [reply doc/source/user/basics.interoperability.rst]: Done.

File: doc/source/user/basics.interoperability.rst [hunk 9]
```diff
@@ -420,3 +516,4 @@ Further reading
 .. _astropy.units: https://docs.astropy.org/en/stable/units/
 .. _pint: https://pint.readthedocs.io/
 .. _unyt: https://unyt.readthedocs.io/
+.. _DLPack: https://dmlc.github.io/dlpack/latest
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply doc/source/user/basics.interoperability.rst]: Done.

File: doc/source/user/basics.interoperability.rst [hunk 10]
```diff
@@ -118,6 +126,26 @@ as the original object and any attributes/behavior it may have had, is lost.
...
 ``__array__()``, see :ref:`basics.dispatch`.
 
+The DLPack Protocol
+~~~~~~~~~~~~~~~~~~~
+
+The :doc:`DLPack <dlpack:index>` protocol defines a memory-layout of
+strided n-dimensional array objects. It offers the following syntax
+for data exchange:
+
+1. A ``from_dlpack(x)`` function, which accepts (array) objects with a
```
Comment (mattip) [inline doc/source/user/basics.interoperability.rst:136]: Could you make this into links (probably my attempt here is wrong)
```suggestion
1. A `numpy.from_dlpack` function, which accepts (array) objects with a
```

**Comment by mattip** [comment]
Could you add a note in doc/release/upcoming_changes? There is a README there which should explain how to add a `new_feature.rst`

**Comment by tirthasheshpatel** [comment]
Added the function to upcoming changes @mattip!

**Review by mattip — APPROVED** [review]

**Comment by tirthasheshpatel** [comment]
The one failure seems unrelated.

**Comment by mattip** [comment]
Thanks @tirthasheshpatel

**Comment by tirthasheshpatel** [comment]
Thanks @mattip @charris @leofang for the reviews!

**Pull request merged**